onnxscript/onnx_types.py

```diff
@@ -36,131 +33,144 @@ def check_shape(shape):
         check_dim(shape)
 
 
-class TensorType:
-    """ONNX Script representation of a tensor type."""
+tensor_type_registry: dict[DType, TensorType] = {}
+_tensor_type_shape_cache: dict[DType, TensorType] = {}
+
+
+class TensorType(ABC):
+    """ONNX Script representation of a tensor type supporting shape annotations.
+
+    A scalar-tensor of rank 0:
+    ::
+
+        tensor: FLOAT
+
+    A tensor of unknown rank:
+    ::
+
+        tensor: FLOAT[...]
+
+    A tensor of rank 2 of unknown dimensions, with symbolic names:
+    ::
+
+        tensor: FLOAT['M', 'N']
+
+    A tensor of rank 2 of known dimensions:
+    ::
 
-    default_instance: Optional["TensorType"] = None
+        tensor: FLOAT[128, 1024]
+    """
+
+    dtype: ClassVar[DType]
+    shape: ClassVar[Optional[ShapeType]]
+
+    def __new__(cls):
+        raise NotImplementedError("TensorTypes cannot be instantiated")
 
-    def __init__(self, dtype, shape: Optional[ShapeType] = None) -> None:
-        self.dtype = dtype
-        self.shape = shape
-        if shape is not None:
+    def __init_subclass__(cls, dtype: DType, shape: Optional[ShapeType] = None):
+        cls.dtype = dtype
+        cls.shape = shape
+        if shape is None:
+            existing_cls = tensor_type_registry.get(dtype)
+            if existing_cls is not None:
+                raise ValueError(
+                    f"Invalid usage: subclass {existing_cls!r} "
+                    f"already defined for dtype={dtype}"
+                )
+            tensor_type_registry[dtype] = cls
+        else:
             check_shape(shape)
 
     def __getitem__(self, shape: Optional[ShapeType]):
-        if self.shape is not None:
+        raise NotImplementedError("should not be reached")
+
+    def __class_getitem__(cls, shape: Optional[ShapeType]) -> TensorType:
```

Comment:
A Python question: does Python overload `TensorType` to also represent its meta-type?

Reply:
I may not be using the right terminology. This returns an instance of `type` class, not an instance of `TensorType` class. Wondering how that is represented in Python? I assume mypy doesn't complain, so trying to understand this

Reply:
Shouldn't the return type be `type` or `Type[TensorType]` or something like that? `Type[TensorType]` doesn't seem quite right, since it returns a derived class, but I am just trying to clarify my question.

Reply:
Oh, got it. Yes, good catch. It should be annotated through `Type[]`. Basically new types are created (and cached for reuse) on demand as shapes are specified through the type.